File: src/test/java/com/tngtech/jgiven/reference/ScenarioStateUsageProviderIntegrationTest.java
package com.tngtech.jgiven.reference;

import com.google.common.collect.Iterables;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.usageView.UsageInfo;
import com.tngtech.jgiven.BaseTestCase;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.*;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

@SuppressWarnings("ConstantConditions")
public class ScenarioStateUsageProviderIntegrationTest extends BaseTestCase {
    public void test_find_Usages_for_type() throws Exception {
        libraryTestUtil.addJGiven();
        configureByFile("ForClass.java");

        Collection<UsageInfo> usages = findUsages();

        assertThat(getOnlyUsageFilename(usages)).contains("SomeOtherReference.java");
    }

    public void test_find_Usages_for_name() throws Exception {
        libraryTestUtil.addJGiven();
        configureByFile("ForName.java");

        Collection<UsageInfo> usages = findUsages();

        assertThat(getOnlyUsageFilename(usages)).contains("SomeReference.java");
    }

    public void test_find_Usages_for_java_lang_reference_resolves_by_name_and_not_by_type() throws Exception {
        libraryTestUtil.addJGiven();
        configureByFile("ForJavaLang.java");

        Collection<UsageInfo> usages = findUsages();

        assertThat(getOnlyUsageFilename(usages)).contains("SomeReference.java");
    }

    public void test_find_no_usages_if_JGiven_is_not_present() throws Exception {
        libraryTestUtil.addJGiven();
        myFixture.configureByFile("NoJGiven.java");

        Collection<UsageInfo> usages = findUsages();

        assertThat(usages).isEmpty();
    }

    @NotNull
    private String getOnlyUsageFilename(Collection<UsageInfo> usages) {
        return Iterables.getOnlyElement(usages).getFile().getName();
    }

    @NotNull
    private Collection<UsageInfo> findUsages() {
        PsiElement element = myFixture.getElementAtCaret();
        return myFixture.findUsages(element);
    }

    @Override
    protected String getTestDataSubDirectory() {
        return "findUsages";
    }

    private void configureByFile(String filename) {
        List<String> files = new ArrayList<>();
        files.add(filename);
        files.addAll(getUsageFilesForAllTests());
        String[] asArray = files.toArray(new String[files.size()]);
        PsiFile[] loadedFiles = myFixture.configureByFiles(asArray);
        assertThat(loadedFiles).hasSameSizeAs(files);
    }

    private Set<String> getUsageFilesForAllTests() {
        return Arrays.stream(new File(getTestDataPath() + "/common").listFiles())
                .map(File::getName)
                .map(name -> "common/" + name)
                .collect(Collectors.toSet());
    }
}

File: src/test/java/com/tngtech/jgiven/reference/ScenarioStateUsageProviderIntegrationTest.kt
package com.tngtech.jgiven.reference

import com.google.common.collect.Iterables
import com.intellij.usageView.UsageInfo
import com.tngtech.jgiven.BaseTestCase
import org.assertj.core.api.Assertions.assertThat
import java.io.File

class ScenarioStateUsageProviderIntegrationTest : BaseTestCase() {

    private val usageFilesForAllTests: Set<String>
        get() = File("$testDataPath/common").listFiles()!!
                .map { it.name }
                .map { name -> "common/$name" }
                .toSet()

    fun test_find_Usages_for_type() {
        configureByFile("ForClass.java")

        val usages = findUsages()

        assertThat(getOnlyUsageFilename(usages)).contains("SomeOtherReference.java")
    }

    fun test_find_Usages_for_name() {
        configureByFile("ForName.java")

        val usages = findUsages()

        assertThat(getOnlyUsageFilename(usages)).contains("SomeReference.java")
    }

    fun test_find_Usages_for_java_lang_reference_resolves_by_name_and_not_by_type() {
        configureByFile("ForJavaLang.java")

        val usages = findUsages()

        assertThat(getOnlyUsageFilename(usages)).contains("SomeReference.java")
    }

    fun test_find_no_usages_if_JGiven_is_not_present() {
        myFixture.configureByFile("NoJGiven.java")

        val usages = findUsages()

        assertThat(usages).isEmpty()
    }

    private fun getOnlyUsageFilename(usages: Collection<UsageInfo>) =
            Iterables.getOnlyElement(usages).file!!.name

    private fun findUsages(): Collection<UsageInfo> {
        val element = myFixture.elementAtCaret
        return myFixture.findUsages(element)
    }

    override val testDataSubDirectory = "findUsages"

    private fun configureByFile(filename: String) {
        val files = listOf(filename) + usageFilesForAllTests
        val loadedFiles = myFixture.configureByFiles(*files.toTypedArray())
        assertThat(loadedFiles).hasSameSizeAs(files)
    }
}